src/Extensions/LanguageConfiguration.re

```diff
@@ -0,0 +1,137 @@
+/*
+ * LanguageConfiguration.re
+ */
+
+open Oni_Core.Utility;
+module Json = Oni_Core.Json;
+
+module AutoClosingPair = {
+  type scopes =
+    | String
+    | Comment
+    | Other(string)
+    | None;
+
+  type t = {
+    openPair: string,
+    closePair: string,
+    notIn: list(scopes),
+  };
+
+  module Decode = {
+    open Json.Decode;
+
+    let scope =
+      string
+      |> map(
+           fun
+           | "string" => String
+           | "comment" => Comment
+           | v => Other(v),
+         );
+
+    let scopeToList = scope |> map(v => [v]);
+
+    // Pulled this from: https://github.com/mattjbray/ocaml-decoders/blob/550c05c16384c9f8d2ab6a36eebc8e1a2684a7e0/src/decode.mli#L106
+    let (>>==::) = (fst, rest) => uncons(rest, fst);
+
+    let tuple =
```

Comment:
This was my first go at using our new JSON decoders - maybe there is a cleaner way @glennsl ? I wanted to parse `["a", "b"]` into `{ openPair: "a", closePair: "b" }`.

I pulled that `>>==::` operator from the docs, but have no idea what it means (perhaps it's a standard nomenclature somewhere).

I think the rest of it was pretty straightforward, though! I like the warm fuzzy feeling of having it covered by the type system... when it compiled, it worked exactly as I expected.

Reply:
Yeah, I'm not a big fan of cryptic symbols, and it definitely doesn't help that `refmt` treats them as linebreaks.

I'd just decode it as a list, then `map` it using a function that matches exactly a two-element array:

```
list |> map(fun | [openPair, closePair] => { openPair, closePair } | _ => failwith("oh no!"))
```

Although it might be more appropriate to use `and_then`, `succeed` and `fail`:

```
list |> and_then(fun
            | [openPair, closePair] => succeed({ openPair, closePair })
            | _ => fail("oh no!")
          );
```

Reply:
Otherwise the decoders look great!

Reply:
Ah, thank you! I was looking for something like `and_then` and couldn't find it (was searching for a `bind`, `flatMap`, etc) - that's exactly what I needed. Much cleaner than what I had before, w/o the cryptic symbols.

Updated, and tests pass still pass 👍

Reply:
`and_then` is what Elm and Rust calls it, because it usually reads better. Not very intuitive when trying to find it though.